### PR TITLE
BUG: Make `arr.ctypes.data` hold onto a reference to the underlying array

### LINKED ISSUE #9647: Segmentation fault when trying to convert numpy array to the C array with ctypes
Python 2.7.12
Numpy 1.13.0

Here is error reproduction:
```
import numpy as np
from ctypes import POINTER, c_float

data =  np.ascontiguousarray(np.array([1]*10000, dtype=np.float32)/255).ctypes.data_as(POINTER(c_float))
print data.contents.value # ok

data =  np.ascontiguousarray(np.array([1]*1000000, dtype=np.float32)/255).ctypes.data_as(POINTER(c_float))
print data.contents.value #giving segmentation fault
```

### PR DESCRIPTION
**Pull request by eric-wieser**
We do this by going through ctypes.from_buffer, which keeps the object the buffer came from alive.

Fixes #9647

### CODE REVIEW

**Comment by eric-wieser** [comment]
Hmm, seems I need to get a 0d array before I request a buffer of it. I think this approach might also fail for arrays containing custom dtypes, or structured types with padding / unions. Perhaps the approach needs to be `pointer = from_buffer(make_writeable(make_0d(arr).view(np.char)))`

File: numpy/core/_internal.py [hunk 1]
```diff
@@ -243,14 +243,36 @@ def cast(self, num, obj):
     def c_void_p(self, num):
         return num
 
+
+def _get_void_ptr(arr):
+    """
+    Get a `ctypes.c_void_p` to arr.data, that keeps a reference to the array
```
Comment (eric-wieser) [inline numpy/core/_internal.py]: I wish `ctypes` would provide a low-level API to save us from doing this dance

**Comment by mattip** [comment]
I am confused by the removal of the WIP label, have we reached the conclusion that this works for 0d arrays and complicated struct dtypes?

**Comment by eric-wieser** [comment]
I think I got it to that state, yes - but clearly I need some more tests to prove that.

**Comment by eric-wieser** [comment]
Tests updated.

**Comment by eric-wieser** [comment]
This is failing the same way on Shippable as #12309...

**Comment by tylerjereddy** [comment]
Except the failure is in Python 3.7, so maybe not a backport issue after all

**Comment by eric-wieser** [comment]
Restarting CI to deal with the shippable failure

**Comment by charris** [comment]
Thanks Eric.

**Comment by charris** [comment]
This seems to have broke azure pipelines. Not sure what has happened, maybe a merge mixup?

**Comment by charris** [comment]
See https://dev.azure.com/numpy/numpy/_build/results?buildId=1265

**Comment by charris** [comment]
Or maybe the test depends on something missing on azure.

**Comment by eric-wieser** [comment]
I think the test got added after this PR was made. I'll take a look.

**Comment by eric-wieser** [comment]
Looks like this patch is to blame, and there's a missing `.view(np.uint8)` needed here

**Comment by charris** [comment]
How do you recommend to deal with this?

**Comment by eric-wieser** [comment]
I have a patch in progress. We could revert in the process

**Comment by charris** [comment]
OK, I'll revert. I'm thinking it would be nice to have all the ctypes fixes and enhancements together in 1.16.

**Comment by charris** [comment]
See #12571, I'll put it in if tests pass.

**Comment by eric-wieser** [comment]
Fixed in #12572, I think.

**Pull request merged**